packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,278 @@
+
+    constructor(
+        private readonly _queue: IDeltaQueue<T>,
+        private readonly _virtualizeValue: (value: T, storeValue?: boolean) => T,
+    ) {
+        super();
+
+        this.on("newListener", (event, _listener) => {
+            assert((this.listeners(event).length !== 0) === this._trackedListeners.has(event), "mismatch");
+            if (!this._trackedListeners.has(event)) {
+                let listener: (...args: any[]) => void;
+
+                // eslint-disable-next-line unicorn/prefer-ternary
+                if (event === "push" || event === "op") {
+                    listener = (task: T) => { this.emit(event, this._virtualizeValue(task, true /* storeValue */)); }
+                } else {
+                    listener = (...args: any[]) => { this.emit(event, ...args); };
+                }
+
+                this.addTrackedListener(event, listener);
+            }
+        });
+    }
+
+    private addTrackedListener(event: string, listener: (...args: any[]) => void) {
+        (this._queue as any).on(event, listener); // TODO: find better cast than "any"
+        assert(!this._trackedListeners.has(event), "double tracked listener");
+        this._trackedListeners.set(event, listener);
+    }
+
+    private removeTrackedListeners() {
+        for (const [event, listener] of this._trackedListeners.entries()) {
+            (this._queue as any).off(event, listener); // TODO: find better cast than "any"
+        }
+        this._trackedListeners.clear();
+    }
+
+    public peek(): T | undefined {
+        const value = this._queue.peek();
+        return value !== undefined ? this._virtualizeValue(value) : undefined;
+    }
+
+    public toArray(): T[] {
+        return this._queue.toArray().map((it) => this._virtualizeValue(it));
+    }
+
+    public dispose(error?: Error | undefined): void {
+        this.removeTrackedListeners();
+        return this._queue.dispose(error);
+    }
+
+    public get paused(): boolean {
+        return this._queue.paused;
+    }
+
+    public get length(): number {
+        return this._queue.length;
+    }
+
+    public get idle(): boolean {
+        return this._queue.idle;
+    }
+
+    public get disposed(): boolean {
+        return this._queue.disposed;
+    }
+
+    public async pause(): Promise<void> {
+        return this._queue.pause();
+    }
+
+    public resume(): void {
+        return this._queue.resume();
+    }
+
+    public async waitTillProcessingDone(): Promise<{ count: number; duration: number; }> {
+        return this._queue.waitTillProcessingDone();
+    }
+}
+
+export class VirtualDeltaManager<TConnectionManager extends IConnectionManager>
+    extends DeltaManager<TConnectionManager> {
+
+    private _virtualSequenceNumber = 0;
+    private _clientSequenceNumber = 0;
+
+    /**
+     * Key: real client sequence number
+     * Value: virtual client sequence number
+     */
+    private readonly _clientSequenceNumberMap = new Map<number, number>();
```

Comment:
Do we need this map? I.e. we talked about stop projecting clientSequenceNubmers to runtime, and then probably we do not need it?